// @squarespace/video-background-rendering/dist/constants/filter.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
var filterOptions = ['none', 'blur', 'brightness', 'contrast', 'invert', 'opacity', 'saturate', 'sepia', 'drop-shadow', 'grayscale', 'hue-rotate'];

/**
 * Each filter style needs to adjust the strength value (1 - 100) by a `modifier`
 * function and a unit, as appropriate. The `modifier` is purely subjective.
 */
var filterProperties = {
  blur: {
    modifier: function modifier(value) {
      return value * 0.3;
    },
    unit: 'px'
  },
  brightness: {
    modifier: function modifier(value) {
      return value * 0.009 + 0.1;
    },
    unit: ''
  },
  contrast: {
    modifier: function modifier(value) {
      return value * 0.4 + 80;
    },
    unit: '%'
  },
  grayscale: {
    modifier: function modifier(value) {
      return value;
    },
    unit: '%'
  },
  'hue-rotate': {
    modifier: function modifier(value) {
      return value * 3.6;
    },
    unit: 'deg'
  },
  invert: {
    modifier: function modifier(value) {
      return 1;
    },
    unit: ''
  },
  opacity: {
    modifier: function modifier(value) {
      return value;
    },
    unit: '%'
  },
  saturate: {
    modifier: function modifier(value) {
      return value * 2;
    },
    unit: '%'
  },
  sepia: {
    modifier: function modifier(value) {
      return value;
    },
    unit: '%'
  }
};

exports.filterOptions = filterOptions;
exports.filterProperties = filterProperties;
